File: hs/src/Day5.hs
{-# OPTIONS_GHC  -Wall #-}

module Day5 where

-- import Data.List (elemIndex)

decodeRowLocator :: String -> Int
decodeRowLocator locator = 
    go locator 0 127
    where
        go :: String -> Int -> Int -> Int
        go [] _ upper = upper
        go (c:cs) lower upper =
            case c of 
                'F' -> 
                    let upper' = (lower + upper) `div` 2 in
                        go cs lower upper'
                'B' -> 
                    let lower' = (lower + upper) `div` 2 + 1 in
                        go cs lower' upper
                _ -> error "Invalid symbol in row locator"

decodeColLocator :: String -> Int
decodeColLocator locator =
    go locator 0 7
    where
        go :: String -> Int -> Int -> Int
        go [] lower upper | lower == upper = upper
                          | otherwise = error "Search error"
        go (c:cs) lower upper =
            case c of 
                'L' -> 
                    let upper' = (lower + upper) `div` 2 in
                        go cs lower upper'
                'R' -> 
                    let lower' = (lower + upper) `div` 2 + 1 in
                        go cs lower' upper
                _ -> error "Invalid symbol in row locator"    

decodeLocator :: String -> Int
decodeLocator locator = 
    let rowLocator = take 7 locator in
    let colLocator = reverse $ take 3 $ reverse locator in
    let row = decodeRowLocator rowLocator in
    let col = decodeColLocator colLocator in
        row * 8 + col

day5 :: IO ()
day5 = do

    locators <- lines <$> readFile "../inputs/Day5.txt"

    let seatIDs = map decodeLocator locators

    let highest = maximum seatIDs

    let mySeat = findMySeat seatIDs

    print highest
    
    print mySeat

    where

        findMySeat ids =
            let lowest = minimum ids in
            let highest = maximum ids in
                head [i | i <- [lowest..highest], i `notElem` ids]
         

File: py/day5.py
#!/usr/bin/python3

def localize_row(row_locator):

    lower = 0
    upper = 127

    for r in row_locator:

        if r == "F":
            upper = (upper + lower) // 2

        elif r == "B":
            lower = (upper + lower) // 2 + 1

    return upper


def localize_column(column_locator):

    lower = 0
    upper = 7

    for r in column_locator:

        if r == "L":
            upper = (upper + lower) // 2

        elif r == "R":
            lower = (upper + lower) // 2 + 1

    return upper


def localize_seat_id(locator):

    row_locator = locator[:7]
    column_locator = locator[-3:]

    row = localize_row(row_locator)
    column = localize_column(column_locator)

    return row * 8 + column


def find_highest_seat_id(occupied_seats):
    return max(occupied_seats)


def find_my_seat(occupied_seats):

    first_seat = min(occupied_seats)
    last_seat = max(occupied_seats)

    for i in range(first_seat, last_seat):
        if i not in occupied_seats:
            return i


def main():

    occupied_seats = [
        localize_seat_id(line.strip())
        for line in open("../inputs/Day5.txt", "r")
    ]

    print(find_highest_seat_id(occupied_seats))
    print(find_my_seat(occupied_seats))


if __name__ == '__main__':
    assert localize_seat_id("FBFBBFFRLR") == 357
    main()
